### PR TITLE
fix guestOsMapper and move mapping code to latest upgrade

### PR DESCRIPTION
**Pull request by DaanHoogland**
### Description

This PR fixes a buh in the guestos mapping facility and moves mappings, so current installations get the right mappings.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [x] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [x] Major
- [ ] Minor
- [ ] Trivial

### How Has This Been Tested?

upgraded from 4.16.0 and checked if the mappings where alright.
upgraded from 4.17 checked that mapping where alright and for the ubuntu on xenserver that an existing VM could still be stopped/started and migrated.

### CODE REVIEW

**Comment by weizhouapache** [comment]
it seems some codes in Upgrade41400to41500 needs to be moved as well. 

```
$ grep -rl addGuestOsAndHypervisorMappings
engine/schema/src/main/java/com/cloud/upgrade/dao/Upgrade41500to41510.java
engine/schema/src/main/java/com/cloud/upgrade/dao/Upgrade41400to41500.java
engine/schema/src/main/java/com/cloud/upgrade/GuestOsMapper.java

$ grep -rl updateGuestOsNameFromMapping
engine/schema/src/main/java/com/cloud/upgrade/dao/Upgrade41500to41510.java
engine/schema/src/main/java/com/cloud/upgrade/GuestOsMapper.java
```

File: engine/schema/src/main/java/com/cloud/upgrade/dao/Upgrade41500to41510.java [hunk 1]
```diff
@@ -250,130 +249,6 @@ private void correctGuestOsNames(final Connection conn) {
         guestOsMapper.updateGuestOsNameFromMapping("OpenSUSE Linux (32 bit)", mapping);
     }
```
Comment (DaanHoogland) [inline engine/schema/src/main/java/com/cloud/upgrade/dao/Upgrade41500to41510.java:250]: @weizhouapache , your comment in https://github.com/apache/cloudstack/pull/7095#issuecomment-1382247322 means that also this method must be moved? It contains a call to `updateGuestOsNameFromMapping()` as well.

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan package

File: engine/schema/src/main/java/com/cloud/upgrade/dao/Upgrade41500to41510.java (hunk 1 above)
Comment (weizhouapache) [reply engine/schema/src/main/java/com/cloud/upgrade/dao/Upgrade41500to41510.java]: double-checked just now.

both `addGuestOsAndHypervisorMappings` and `updateGuestOsIdInHypervisorMapping` call the method `getGuestOsId`, they need to be moved.

`updateGuestOsNameFromMapping` is correct. sorry for my mistake

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Review by nvazquez — DISMISSED** [review]
LGTM, haven't tested it

File: engine/schema/src/main/java/com/cloud/upgrade/dao/Upgrade41500to41510.java [hunk 2]
```diff
@@ -69,8 +66,6 @@ public InputStream[] getPrepareScripts() {
 
     @Override
     public void performDataMigration(Connection conn) {
-        correctGuestOsNames(conn);
-        updateGuestOsMappings(conn);
```
Comment (nvazquez) [inline engine/schema/src/main/java/com/cloud/upgrade/dao/Upgrade41500to41510.java:73]: I'm not sure about these schema changes, what would happen to templates/VMs created from existing guest OS on previous versions?

**Review by weizhouapache — APPROVED** [review]
code lgtm

we need to verify the guest os mapping in environments below
(1) fresh 4.18
(2) upgrade from 4.17 to 4.18

I will verify when packages are done.

**Comment by weizhouapache** [comment]
@blueorangutan package

**Review by harikrishna-patnala — APPROVED** [review]
Code LGTM.

**Comment by weizhouapache** [comment]
> LGTM @DaanHoogland would we need to do that with every release?

if we will do another 4.17 minor release (4.17.3.0), we need to fix in 4.17 as well.

**Comment by DaanHoogland** [comment]
@weizhouapache should I rebase on 4.17?

**Comment by DaanHoogland** [comment]
> LGTM @DaanHoogland would we need to do that with every release?

Basically, no. Only because the mapping code was broken between 4.16.1 and 4.18 we need to redo the mapping in 4.18

**Comment by weizhouapache** [comment]
> @weizhouapache should I rebase on 4.17?

it will be complicated ...

you need to 
(1) add upgrade path 4.17.2.0->4.17.3.0
(2) move everything to Upgrade41720to41730.java

and
what if there is not 4.17.3.0 release ? 
what if 4.17.3.0 is released after 4.18.0.0, can upgrade from 4.17.3.0 to 4.18.0.0 ?
we may need to keep the codes in both files: Upgrade41720to41730.java and Upgrade41720to41800.java

**Comment by DaanHoogland** [comment]
> > @weizhouapache should I rebase on 4.17?
> 
> @DaanHoogland 
> 
> it will be complicated ...
> 
> you need to 
> (1) add upgrade path 4.17.2.0->4.17.3.0
> (2) move everything to Upgrade41720to41730.java
> 
> and
> what if there is not 4.17.3.0 release ? 
> what if 4.17.3.0 is released after 4.18.0.0, can upgrade from 4.17.3.0 to 4.18.0.0 ?
> we may need to keep the codes in both files: Upgrade41720to41730.java and Upgrade41720to41800.java

And merge forward guaranteed conflicts. Let's leave it as is.

**Comment by vladimirpetrov** [comment]
@blueorangutan package

**Review by vladimirpetrov — APPROVED** [review]
LGTM based on manual testing.

**Comment by BryanMLima** [comment]
Hey @DaanHoogland, I know it has been a while, but did you check if the Guest OS types were duplicate after this change? I have some environments that were recently updated to the `4.18` from the `4.16` and there are 43 duplicate Guest OS types.

**Comment by weizhouapache** [comment]
@BryanMLima 
yes, I have noticed it. 
The method `addGuestOsAndHypervisorMappings` used in this PR will add a new guest OS (even it already exists) and then add guest os mappings.
I have created an issue to track it: #7783

**Pull request merged**